WireAPI/Sources/WireAPI/Public/Provider/APIProvider.swift

```diff
@@ -0,0 +1,50 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+/// A type responsible for providing apis to consumers.
+
+public struct APIProvider {
+
+    public let httpClient: HTTPClient
+
+    public init(httpClient: HTTPClient) {
+        self.httpClient = httpClient
+    }
+
+    public func backendInfoAPI(for version: APIVersion) -> BackendInfoAPI {
+        switch version {
+        case .v0:
+            BackendInfoAPIV0(httpClient: httpClient)
+        case .v1:
+            BackendInfoAPIV1(httpClient: httpClient)
+        case .v2:
+            BackendInfoAPIV2(httpClient: httpClient)
+        case .v3:
+            BackendInfoAPIV3(httpClient: httpClient)
+        case .v4:
+            BackendInfoAPIV4(httpClient: httpClient)
+        case .v5:
+            BackendInfoAPIV5(httpClient: httpClient)
+        case .v6:
+            BackendInfoAPIV6(httpClient: httpClient)
+        }
+    }
```

Comment:
Using the `APIProvider` in such way centralises all APIs in one place. When I wrote the example of an `APIService` it was critiqued it not scaling well.

Isn't it here the same? Can we have a simple object to keep it seperated?
```swift
public struct BackendInfoAPIClient: BackendInfoAPI {
    let httpClient: httpClient

    // maybe even use the `version` as property over parameters.
    public func backendInfoAPI(for version: APIVersion) -> BackendInfoAPI { ... }
}
```

Reply:
Yep you're right we can split it up. I'm confused though by the example, is `BackendInfoAPIClient` the client that conforms to `BackendInfoAPI` or is it a provider object that provides a `BackendInfoAPI` object?

Another idea that comes to mind is to create a static function on the protocol to provide the implementation, for instance: 

```swift
extension BackendInfoAPI {

    public static func apiClient(
        for version: APIVersion,
        httpClient: HTTPClient
    ) -> any BackendInfoAPI {
        switch version {
        case .v0:
            BackendInfoAPIV0(httpClient: httpClient)
        case .v1:
            BackendInfoAPIV1(httpClient: httpClient)
        case .v2:
            BackendInfoAPIV2(httpClient: httpClient)
        case .v3:
            BackendInfoAPIV3(httpClient: httpClient)
        case .v4:
            BackendInfoAPIV4(httpClient: httpClient)
        case .v5:
            BackendInfoAPIV5(httpClient: httpClient)
        case .v6:
            BackendInfoAPIV6(httpClient: httpClient)
        }
    }

}
```

Reply:
done